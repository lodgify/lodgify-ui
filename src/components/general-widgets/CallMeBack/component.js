import React from 'react';
import PropTypes from 'prop-types';

import {
  CALL_ME_BACK,
  DATE,
  EMAIL,
  NAME,
  NOTES,
  PHONE,
  PROPERTY,
  SEND,
  TIME,
} from 'utils/default-strings';
import { Dropdown } from 'inputs/Dropdown';
import { Form } from 'collections/Form';
import { ICON_NAMES } from 'elements/Icon';
import { InputGroup } from 'collections/InputGroup';
import { PhoneInput } from 'inputs/PhoneInput';
import { SingleDatePicker } from 'inputs/SingleDatePicker';
import { TextArea } from 'inputs/TextArea';
import { TextInput } from 'inputs/TextInput';

/**
 * The standard widget for a user to request a call back.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({
  dateInputPlaceholder,
  emailInputLabel,
  errorMessage,
  headingText,
  nameInputLabel,
  notesInputLabel,
  onSubmit,
  phoneInputLabel,
  propertyInputLabel,
  propertyOptions,
  submitButtonText,
  successMessage,
  timeInputLabel,
  timeOptions,
  validation,
}) => (
  <Form
    autoComplete="off"
    errorMessage={errorMessage}
    headingText={headingText}
    onSubmit={onSubmit}
    submitButtonText={submitButtonText}
    successMessage={successMessage}
    validation={validation}
  >
    <InputGroup>
      <TextInput autoComplete="name" label={nameInputLabel} name="name" />
      <PhoneInput autoComplete="tel" label={phoneInputLabel} name="phone" />
    </InputGroup>
    <TextInput autoComplete="email" label={emailInputLabel} name="email" />
    <InputGroup>
      <SingleDatePicker name="date" placeholderText={dateInputPlaceholder} />
      <Dropdown
        icon={ICON_NAMES.CLOCK}
        label={timeInputLabel}
        name="time"
        options={timeOptions}
      />
    </InputGroup>
    <Dropdown
      label={propertyInputLabel}
      name="property"
      options={propertyOptions}
    />
    <TextArea label={notesInputLabel} name="notes" />
  </Form>
);

Component.displayName = 'CallMeBack';

Component.defaultProps = {
  dateInputPlaceholder: DATE,
  emailInputLabel: EMAIL,
  errorMessage: '',
  headingText: CALL_ME_BACK,
  nameInputLabel: NAME,
  notesInputLabel: NOTES,
  onSubmit: Function.prototype,
  phoneInputLabel: PHONE,
  propertyInputLabel: PROPERTY,
  submitButtonText: SEND,
  successMessage: '',
  timeInputLabel: TIME,
  validation: {},
};

Component.propTypes = {
  /** The placeholder for the date input. */
  dateInputPlaceholder: PropTypes.string,
  /** The label for the email input. */
  emailInputLabel: PropTypes.string,
  /** The message to display when the form has an error. */
  errorMessage: PropTypes.string,
  /** The text to display as a heading at the top of the widget. */
  headingText: PropTypes.string,
  /** The label for the name input. */
  nameInputLabel: PropTypes.string,
  /** The label for the notes input. */
  notesInputLabel: PropTypes.string,
  /** The function to call when the form is submitted
   *  @param {Object} values - The values of the inputs in the form.
   */
  onSubmit: PropTypes.func,
  /** The label for the phone input */
  phoneInputLabel: PropTypes.string,
  /** The label for the property input. */
  propertyInputLabel: PropTypes.string,
  /** The options which the user can select for the property field. */
  propertyOptions: PropTypes.arrayOf(
    PropTypes.shape({
      /** The visible text for the option. */
      text: PropTypes.string.isRequired,
      /** The underlying value for the option. */
      value: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.number,
        PropTypes.string,
      ]),
    })
  ).isRequired,
  /** The form submit button text */
  submitButtonText: PropTypes.string,
  /** The message to display when the form is successful. */
  successMessage: PropTypes.string,
  /** The label for the time input. */
  timeInputLabel: PropTypes.string,
  /** The options which the user can select for the time field. */
  timeOptions: PropTypes.arrayOf(
    PropTypes.shape({
      /** The visible text for the option. */
      text: PropTypes.string.isRequired,
      /** The underlying value for the option. */
      value: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.number,
        PropTypes.string,
      ]),
    })
  ).isRequired,
  /** Settings for validating inputs. Each value should match [the shape documented in `Form`](https://lodgify.github.io/lodgify-ui/#/Collections/Form) */
  validation: PropTypes.shape({
    date: PropTypes.object,
    email: PropTypes.object,
    name: PropTypes.object,
    notes: PropTypes.object,
    phone: PropTypes.object,
    property: PropTypes.object,
    time: PropTypes.object,
  }),
};
