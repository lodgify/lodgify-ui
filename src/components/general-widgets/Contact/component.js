import React from 'react';
import PropTypes from 'prop-types';

import { Form } from 'collections/Form';
import { InputGroup } from 'collections/InputGroup';
import { TextInput } from 'inputs/TextInput';
import { PhoneInput } from 'inputs/PhoneInput';
import { NumberInput } from 'inputs/NumberInput';
import { DateRangePicker } from 'inputs/DateRangePicker';
import { TextArea } from 'inputs/TextArea';
import { CaptchaInput } from 'inputs/CaptchaInput';
import { Dropdown } from 'inputs/Dropdown';
import {
  ARRIVAL,
  CONTACT,
  COMMENTS,
  DEPARTURE,
  EMAIL,
  GUESTS,
  NAME,
  PHONE,
  PROPERTY,
  ROOM,
  SECURITY_CODE,
  SEND,
} from 'utils/default-strings';

/**
 * The standard widget for a user contact an owner.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({
  arrivalDateInputLabel,
  captchaInputImage,
  captchaInputLabel,
  commentsInputLabel,
  dateRangePickerLocaleCode,
  departureDateInputLabel,
  emailInputLabel,
  errorMessage,
  guestsInputLabel,
  headingText,
  nameInputLabel,
  onChangeProperty,
  onSubmit,
  phoneInputLabel,
  propertyInputLabel,
  propertyOptions,
  roomInputLabel,
  roomOptions,
  submitButtonText,
  successMessage,
  validation,
}) => (
  <Form
    autoComplete="off"
    errorMessage={errorMessage}
    headingText={headingText}
    onSubmit={onSubmit}
    submitButtonText={submitButtonText}
    successMessage={successMessage}
    validation={validation}
  >
    <InputGroup>
      <TextInput autoComplete="name" label={nameInputLabel} name="name" />
      <PhoneInput autoComplete="tel" label={phoneInputLabel} name="phone" />
    </InputGroup>
    <TextInput autoComplete="email" label={emailInputLabel} name="email" />
    <InputGroup>
      <DateRangePicker
        endDatePlaceholderText={departureDateInputLabel}
        localeCode={dateRangePickerLocaleCode}
        name="dates"
        startDatePlaceholderText={arrivalDateInputLabel}
        width="eight"
      />
      <NumberInput
        label={guestsInputLabel}
        min={1}
        name="guests"
        width="four"
      />
    </InputGroup>
    <TextArea label={commentsInputLabel} name="comments" />
    {(roomOptions || propertyOptions) && (
      <InputGroup>
        {propertyOptions && (
          <Dropdown
            label={propertyInputLabel}
            name="property"
            onChange={onChangeProperty}
            options={propertyOptions}
          />
        )}
        {roomOptions && (
          <Dropdown label={roomInputLabel} name="room" options={roomOptions} />
        )}
      </InputGroup>
    )}
    <CaptchaInput
      image={captchaInputImage}
      label={captchaInputLabel}
      name="captcha"
    />
  </Form>
);

Component.displayName = 'Contact';

Component.defaultProps = {
  arrivalDateInputLabel: ARRIVAL,
  captchaInputLabel: SECURITY_CODE,
  commentsInputLabel: COMMENTS,
  dateRangePickerLocaleCode: undefined,
  departureDateInputLabel: DEPARTURE,
  emailInputLabel: EMAIL,
  errorMessage: '',
  guestsInputLabel: GUESTS,
  headingText: CONTACT,
  nameInputLabel: NAME,
  onChangeProperty: Function.prototype,
  onSubmit: Function.prototype,
  phoneInputLabel: PHONE,
  propertyInputLabel: PROPERTY,
  propertyOptions: null,
  roomInputLabel: ROOM,
  roomOptions: null,
  submitButtonText: SEND,
  successMessage: '',
  validation: {},
};

Component.propTypes = {
  /** The label for the arrival date input.*/
  arrivalDateInputLabel: PropTypes.string,
  /** The source url for the image to display. */
  captchaInputImage: PropTypes.string.isRequired,
  /** The label for the captcha input.*/
  captchaInputLabel: PropTypes.string,
  /** The label for the comments input.*/
  commentsInputLabel: PropTypes.string,
  /** The ISO 639-1 locale code which changes the format and language of days of the week and the months of the year in the date range picker. */
  dateRangePickerLocaleCode: PropTypes.string,
  /** The label for the departure date input.*/
  departureDateInputLabel: PropTypes.string,
  /** The label for the email input.*/
  emailInputLabel: PropTypes.string,
  /** The message to display when the form has an error. */
  errorMessage: PropTypes.string,
  /** The label for the guests input.*/
  guestsInputLabel: PropTypes.string,
  /** The text to display as a heading at the top of the widget. */
  headingText: PropTypes.string,
  /** The label for the name input.*/
  nameInputLabel: PropTypes.string,
  /** The function called when the property dropdown is changed.
   *  @param {String}        name - The name of the property dropdown field.
   *  @param {String|Number} value - The value of the property dropdown after the change
   */
  onChangeProperty: PropTypes.func,
  /** The function to call when the form is submitted
   *  @param {Object} values - The values of the inputs in the form.
   */
  onSubmit: PropTypes.func,
  /** The label for the phone input.*/
  phoneInputLabel: PropTypes.string,
  /** The label for the property input.*/
  propertyInputLabel: PropTypes.string,
  /** The options which the user can select for the property field. */
  propertyOptions: PropTypes.arrayOf(
    PropTypes.shape({
      /** The indent level of a property option. One of: 0, 1, 2, 3, 4, 5 */
      indent: PropTypes.oneOf([0, 1, 2, 3, 4, 5]),
      /** The visible text for the option. */
      text: PropTypes.string.isRequired,
      /** The underlying value for the option. */
      value: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.number,
        PropTypes.string,
      ]),
    })
  ),
  /** The label for the room input.*/
  roomInputLabel: PropTypes.string,
  /** The options which the user can select for the room field. */
  roomOptions: PropTypes.arrayOf(
    PropTypes.shape({
      /** The visible text for the option. */
      text: PropTypes.string.isRequired,
      /** The underlying value for the option. */
      value: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.number,
        PropTypes.string,
      ]),
    })
  ),
  /** The text to display on the submit button. */
  submitButtonText: PropTypes.string,
  /** The message to display when the form is successful. */
  successMessage: PropTypes.string,
  /** Settings for validating inputs. Each value should match [the shape documented in `Form`](https://lodgify.github.io/lodgify-ui/#/Collections/Form) */
  validation: PropTypes.shape({
    captcha: PropTypes.object,
    comments: PropTypes.object,
    dates: PropTypes.object,
    email: PropTypes.object,
    guests: PropTypes.object,
    name: PropTypes.object,
    phone: PropTypes.object,
    property: PropTypes.object,
    room: PropTypes.object,
  }),
};
