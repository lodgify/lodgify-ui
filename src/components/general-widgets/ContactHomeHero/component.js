import React from 'react';
import {
  string,
  bool,
  oneOf,
  func,
  shape,
  object,
  number,
  oneOfType,
  node,
  arrayOf,
} from 'prop-types';

import { FullBleed } from 'media/FullBleed';
import { ContactHeader } from 'collections/ContactHeader';
import { FlexContainer } from 'layout/FlexContainer';
import { Grid } from 'layout/Grid';
import { GridRow } from 'layout/GridRow';
import { Heading } from 'typography/Heading';
import { HorizontalGutters } from 'layout/HorizontalGutters';
import { SearchBar } from 'general-widgets/SearchBar';
import { ShowOn } from 'layout/ShowOn';

import { DEFAULT_BOTTOM_OFFSET } from './constants';

/**
 * A hero displays a header and optional children with a full bleed image background
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({
  backgroundImageHeight,
  backgroundImageSizes,
  backgroundImageSrcSet,
  backgroundImageUrl,
  backgroundImageWidth,
  bottomOffset,
  callToActionText,
  className,
  currencyNoResultsText,
  currencyOptions,
  currencyValue,
  headingHref,
  headingText,
  isSearchBarDateRangePickerLoading,
  languageOptions,
  languageValue,
  navigationMenuItems,
  phoneNumber,
  placeholderBackgroundImageUrl,
  primaryButtonText,
  searchBarDateRangePickerLocaleCode,
  searchBarDatesCheckInLabel,
  searchBarDatesCheckOutLabel,
  searchBarDatesInputFocusedInput,
  searchBarDatesInputOnFocusChange,
  searchBarDatesInputValue,
  searchBarGetIsDayBlocked,
  searchBarGuestsInputLabel,
  searchBarGuestsInputValue,
  searchBarLocationInputLabel,
  searchBarLocationInputValue,
  searchBarLocationOptions,
  searchBarMaximumGuestsInputValue,
  searchBarModalHeadingText,
  searchBarOnChangeInput,
  searchBarOnSubmit,
  searchBarSearchButton,
}) => {
  const searchBarSharedProps = {
    dateRangePickerLocaleCode: searchBarDateRangePickerLocaleCode,
    datesCheckInLabel: searchBarDatesCheckInLabel,
    datesCheckOutLabel: searchBarDatesCheckOutLabel,
    datesInputFocusedInput: searchBarDatesInputFocusedInput,
    datesInputOnFocusChange: searchBarDatesInputOnFocusChange,
    datesInputValue: searchBarDatesInputValue,
    getIsDayBlocked: searchBarGetIsDayBlocked,
    guestsInputLabel: searchBarGuestsInputLabel,
    guestsInputValue: searchBarGuestsInputValue,
    locationInputLabel: searchBarLocationInputLabel,
    locationInputValue: searchBarLocationInputValue,
    locationOptions: searchBarLocationOptions,
    maximumGuestsInputValue: searchBarMaximumGuestsInputValue,
    onChangeInput: searchBarOnChangeInput,
    onSubmit: searchBarOnSubmit,
    searchButton: searchBarSearchButton,
    isDateRangePickerLoading: isSearchBarDateRangePickerLoading,
  };

  return (
    <FullBleed
      bottomOffset={bottomOffset}
      className={className}
      hasGradient
      imageHeight={backgroundImageHeight}
      imageUrl={backgroundImageUrl}
      imageWidth={backgroundImageWidth}
      placeholderImageUrl={placeholderBackgroundImageUrl}
      sizes={backgroundImageSizes}
      srcSet={backgroundImageSrcSet}
    >
      <ContactHeader
        className={className}
        currencyNoResultsText={currencyNoResultsText}
        currencyOptions={currencyOptions}
        currencyValue={currencyValue}
        headingHref={headingHref}
        headingText={headingText}
        isBackgroundFilled
        languageOptions={languageOptions}
        languageValue={languageValue}
        navigationMenuItems={navigationMenuItems}
        phoneNumber={phoneNumber}
        primaryButtonText={primaryButtonText}
      />
      <FlexContainer
        alignItems="center"
        flexDirection="column"
        justifyContent="space-evenly"
      >
        <HorizontalGutters>
          <ShowOn
            computer
            parent={Heading}
            parentProps={{
              isColorInverted: true,
              size: 'huge',
              textAlign: 'center',
            }}
            tablet
            widescreen
          >
            {callToActionText}
          </ShowOn>
          <ShowOn
            mobile
            parent={Heading}
            parentProps={{
              isColorInverted: true,
              size: 'large',
              textAlign: 'center',
            }}
          >
            {callToActionText}
          </ShowOn>
        </HorizontalGutters>
        <HorizontalGutters>
          <Grid areColumnsCentered>
            <GridRow horizontalAlignContent="center">
              <ShowOn
                computer
                parent={SearchBar}
                parentProps={{
                  ...searchBarSharedProps,
                  willLocationDropdownOpenAbove: true,
                }}
                tablet
                widescreen
              />
              <ShowOn mobile>
                <SearchBar
                  {...searchBarSharedProps}
                  isDisplayedAsModal
                  modalHeadingText={searchBarModalHeadingText}
                />
              </ShowOn>
            </GridRow>
          </Grid>
        </HorizontalGutters>
      </FlexContainer>
    </FullBleed>
  );
};

Component.displayName = 'ContactHomeHero';

Component.defaultProps = {
  backgroundImageHeight: undefined,
  backgroundImageSizes: undefined,
  backgroundImageSrcSet: undefined,
  backgroundImageWidth: undefined,
  bottomOffset: DEFAULT_BOTTOM_OFFSET,
  callToActionText: '',
  className: null,
  currencyNoResultsText: '',
  currencyOptions: [],
  currencyValue: null,
  headingHref: '',
  headingText: '',
  isHeaderBackgroundFilled: false,
  isSearchBarDateRangePickerLoading: undefined,
  languageOptions: [],
  languageValue: null,
  onChangeCurrency: undefined,
  onChangeLanguage: undefined,
  phoneNumber: '',
  placeholderBackgroundImageUrl: null,
  primaryButtonText: '',
  searchBarDateRangePickerLocaleCode: undefined,
  searchBarDatesCheckInLabel: undefined,
  searchBarDatesCheckOutLabel: undefined,
  searchBarDatesInputFocusedInput: undefined,
  searchBarDatesInputOnFocusChange: Function.prototype,
  searchBarDatesInputValue: null,
  searchBarGetIsDayBlocked: undefined,
  searchBarGuestsInputLabel: undefined,
  searchBarGuestsInputValue: null,
  searchBarLocationInputLabel: undefined,
  searchBarLocationInputValue: undefined,
  searchBarLocationOptions: undefined,
  searchBarMaximumGuestsInputValue: undefined,
  searchBarModalHeadingText: undefined,
  searchBarOnChangeInput: undefined,
  searchBarOnSubmit: Function.prototype,
  searchBarSearchButton: undefined,
};

Component.propTypes = {
  /** The natural height of the background image. */
  backgroundImageHeight: oneOfType([string, number]),
  /** A list of one or more strings separated by commas indicating a set of source sizes for the background image. See [the MDN docs for more information](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img). */
  backgroundImageSizes: string,
  /** A list of one or more strings separated by commas indicating a set of possible image sources for the user agent to use for the background image. See [the MDN docs for more information](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img). */
  backgroundImageSrcSet: string,
  /** The source url of the hero's background image. */
  backgroundImageUrl: string.isRequired,
  /** The natural width of the background image. */
  backgroundImageWidth: oneOfType([string, number]),
  /** Reduce the height of the Hero with an offset, supports CSS dimensions. */
  bottomOffset: string,
  /** The text to display at the center of the hero image. */
  callToActionText: string,
  /** The custom classes. */
  className: string,
  /** The text to display when no results are returned from the currency dropdown. */
  currencyNoResultsText: string,
  /** The options which the user can select for the currency dropdown. */
  currencyOptions: arrayOf(
    shape({
      /** The visible text for the option. */
      text: string.isRequired,
      /** The underlying value for the option. */
      value: oneOfType([bool, number, string]),
    })
  ),
  /** The current value of the currency dropdown. */
  currencyValue: oneOfType([bool, number, string]),
  /** The link for the heading. */
  headingHref: string,
  /** The text for the heading. */
  headingText: string,
  /** The background of the header is a solid color. */
  isHeaderBackgroundFilled: bool,
  /** Is the search bar date range picker in a state of loading. */
  isSearchBarDateRangePickerLoading: bool,
  /** The options which the user can select for the language dropdown. */
  languageOptions: arrayOf(
    shape({
      /** The visible text for the option. */
      text: string.isRequired,
      /** The underlying value for the option. */
      value: oneOfType([bool, number, string]),
    })
  ),
  /** The current value of the language dropdown. */
  languageValue: oneOfType([bool, number, string]),
  /** The items for a user to navigate the site. */
  navigationMenuItems: arrayOf(
    shape({
      /** The href url for an item which is a link. */
      href: string,
      /** Sub navigation items to display as a [`<Submenu />`](#submenu) under an item. */
      subItems: arrayOf(
        shape({
          href: string.isRequired,
          target: string,
          text: string.isRequired,
        })
      ),
      /** Specifies where to display the linked navigation items URL. See [MDN docs `<a />` for more](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-target). */
      target: string,
      /** The visible text for an item. */
      text: string.isRequired,
    })
  ).isRequired,
  /** The function called when the currency dropdown is changed.
   *  @param {String} name
   *  @param {String} value
   */
  onChangeCurrency: func,
  /** The function called when the language dropdown is changed.
   *  @param {String} name
   *  @param {String} value
   */
  onChangeLanguage: func,
  /** The phone number to display. */
  phoneNumber: string,
  /** The ISO 639-1 locale code which changes the format and language of days of the week and the months of the year in the search bars date range picker. */
  placeholderBackgroundImageUrl: string,
  /** The text to display for the primary button. */
  primaryButtonText: string,
  /** The label for the date picker check in input of the search bar when there is no value selected. */
  searchBarDateRangePickerLocaleCode: string,
  /** The label for the date picker check out input of the search bar when there is no value selected. */
  searchBarDatesCheckInLabel: string,
  /** The field of the dates input which is currently focused. Used when consuming `PropertyPageSearchBar` as a controlled component. */
  searchBarDatesCheckOutLabel: string,
  /**
   * A function called when the focus state of the dates input in the search bar changes.
   * @param {String} inputName
   */
  searchBarDatesInputFocusedInput: oneOf([null, 'startDate', 'endDate']),
  /** The value for the dates input of the search bar. Used when consuming `HomepageHero` as a controlled component. */
  searchBarDatesInputOnFocusChange: func,
  /**
   * A function called for each day to be displayed in the DateRangePicker.
   * Returning true blocks that day in the date range picker.
   * @param   {Moment}  day - The day to test.
   * @returns {boolean}     - Is the day blocked.
   */
  searchBarDatesInputValue: shape({
    endDate: object,
    startDate: object,
  }),
  /** The label for the guests input of the search bar. */
  searchBarGetIsDayBlocked: func,
  /** The value for the guests input of the search bar. Used when consuming `HomepageHero` as a controlled component. */
  searchBarGuestsInputLabel: string,
  /** The label for the location input of the search bar when there is no option selected. */
  searchBarGuestsInputValue: number,
  /** The value for the location input of the search bar. Used when consuming `HomepageHero` as a controlled component. */
  searchBarLocationInputLabel: string,
  /** The options which the user can select in the location field of the search bar. */
  searchBarLocationInputValue: string,
  /** The maximum number the guests input counter can increase to. */
  searchBarLocationOptions: arrayOf(
    shape({
      /** The indent level of a location option. One of: 0, 1, 2, 3, 4, 5 */
      indent: oneOf([0, 1, 2, 3, 4, 5]),
      /** The visible text for the option. */
      text: string.isRequired,
      /** The underlying value for the option. */
      value: oneOfType([bool, number, string]),
    })
  ),
  /* The heading displayed in the search bar modal. */
  searchBarMaximumGuestsInputValue: number,
  /** A function called when a change in an input occurs in the search bar.
   *  @param {Object} values - The values of the inputs in the search bar.
   *  @param {Object} values.dates
   *  @param {String} values.guests
   *  @param {String} values.location
   */
  searchBarModalHeadingText: string,
  /** The function to call when the search bar is submitted.
   *  @param {Object} values - The values of the inputs in the search bar.
   *  @param {Object} values.dates
   *  @param {String} values.guests
   *  @param {String} values.location
   */
  searchBarOnChangeInput: func,
  /** The search button the search bar displays. */
  searchBarOnSubmit: func,
  /** The background placeholder image url of the hero. */
  searchBarSearchButton: node,
};
