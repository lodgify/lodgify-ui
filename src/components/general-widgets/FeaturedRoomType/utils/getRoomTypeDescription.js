/**
 * @typedef  {Object} null
 *  @param   {string} guestsLabel
 *  @param   {number} guestsNumber
 *  @param   {string} bedsLabel
 *  @param   {number} bedsNumber
 *  @returns {string|null}
 */
export const getRoomTypeDescription = (
  guestsLabel,
  guestsNumber,
  bedsLabel,
  bedsNumber
) => {
  if (bedsNumber !== undefined && guestsNumber !== undefined) {
    return `${guestsLabel}: ${guestsNumber} | ${bedsLabel}: ${bedsNumber}`;
  }

  if (bedsNumber !== undefined) {
    return `${bedsLabel}: ${bedsNumber}`;
  }

  if (guestsNumber !== undefined) {
    return `${guestsLabel}: ${guestsNumber}`;
  }

  return null;
};
