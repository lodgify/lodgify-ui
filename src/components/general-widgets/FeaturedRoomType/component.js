import React from 'react';
import PropTypes from 'prop-types';
import { Card } from 'semantic-ui-react';

import { ResponsiveImage } from 'media/ResponsiveImage';
import { getNightPriceMarkup } from 'utils/get-night-price-markup/';

import { getRoomTypeDescription } from './utils/getRoomTypeDescription';

/**
 * The standard widget for featuring a room.
 */
// eslint-disable-next-line jsdoc/require-jsdoc
export const Component = ({
  bedsLabel,
  bedsNumber,
  guestsLabel,
  guestsNumber,
  imageAlternativeText,
  imageSizes,
  imageSrcSet,
  imageUrl,
  locationName,
  nightPrice,
  placeholderImageUrl,
  roomTypeName,
  roomTypeUrl,
}) => (
  <Card className="has-featured" href={roomTypeUrl}>
    <ResponsiveImage
      alternativeText={imageAlternativeText}
      imageUrl={imageUrl}
      isFluid
      placeholderImageUrl={placeholderImageUrl}
      sizes={imageSizes}
      srcSet={imageSrcSet}
    />
    <Card.Content>
      <Card.Header>{roomTypeName}</Card.Header>
      <Card.Description>{locationName}</Card.Description>
      <Card.Description>
        {getRoomTypeDescription(
          guestsLabel,
          guestsNumber,
          bedsLabel,
          bedsNumber
        )}
      </Card.Description>
      <Card.Description>{getNightPriceMarkup(nightPrice)}</Card.Description>
    </Card.Content>
  </Card>
);

Component.displayName = 'FeaturedRoomType';

Component.defaultProps = {
  bedsLabel: 'Beds',
  guestsLabel: 'Guests',
  imageAlternativeText: undefined,
  imageSizes: undefined,
  imageSrcSet: undefined,
  placeholderImageUrl: undefined,
};

Component.propTypes = {
  /** The label displayed next to the number of beds. */
  bedsLabel: PropTypes.string,
  /** The number of available beds in the room. */
  bedsNumber: PropTypes.number.isRequired,
  /** The label displayed next to the number of guests. */
  guestsLabel: PropTypes.string,
  /** The number of guests the room can accommodate. */
  guestsNumber: PropTypes.number.isRequired,
  /** The alternative text for the image to display. */
  imageAlternativeText: PropTypes.string,
  /** A list of one or more strings separated by commas indicating a set of source sizes. See [the MDN docs for more information](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img). */
  imageSizes: PropTypes.string,
  /** A list of one or more strings separated by commas indicating a set of possible image sources for the user agent to use. See [the MDN docs for more information](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img). */
  imageSrcSet: PropTypes.string,
  /** URL pointing to the image to display. */
  imageUrl: PropTypes.string.isRequired,
  /** The name of the location of the room. */
  locationName: PropTypes.string.isRequired,
  /** The price per night of the room, with currency symbol. */
  nightPrice: PropTypes.string.isRequired,
  /** URL pointing to the placeholder image to render. */
  placeholderImageUrl: PropTypes.string,
  /** The name of the room. */
  roomTypeName: PropTypes.string.isRequired,
  /** URL pointing to a page with details of the room. */
  roomTypeUrl: PropTypes.string.isRequired,
};
