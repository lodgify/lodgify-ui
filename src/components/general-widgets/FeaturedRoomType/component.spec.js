import React from 'react';
import { shallow } from 'enzyme';
import { Card, Image } from 'semantic-ui-react';
import {
  expectComponentToBe,
  expectComponentToHaveChildren,
  expectComponentToHaveProps,
  expectComponentToHaveDisplayName,
} from '@lodgify/enzyme-jest-expect-helpers';

import { Heading } from 'typography/Heading';

import { getRoomTypeDescription } from './utils/getRoomTypeDescription';
import { Component as FeaturedRoomType } from './component';

const props = {
  bedsNumber: 3,
  guestsNumber: 3,
  imageAlternativeText: 'this alt tag',
  imageUrl: '🐱🐱',
  locationName: 'Catania',
  nightPrice: '$280',
  roomTypeName: 'The Cat House',
  roomTypeUrl: '/',
};

const getFeaturedRoomType = () => shallow(<FeaturedRoomType {...props} />);

describe('<FeaturedRoomType />', () => {
  it('should render a single Semantic UI `Card` component', () => {
    const wrapper = getFeaturedRoomType();

    expectComponentToBe(wrapper, Card);
  });

  describe('the `Card` component', () => {
    const getCard = () => getFeaturedRoomType().find(Card);

    it('should have the right props', () => {
      const wrapper = getCard();

      expectComponentToHaveProps(wrapper, { href: props.roomTypeUrl });
    });

    it('should render the right children', () => {
      const wrapper = getCard();

      expectComponentToHaveChildren(wrapper, Image, Card.Content);
    });
  });

  describe('the `Image` component', () => {
    it('should have the right props', () => {
      const wrapper = getFeaturedRoomType().find(Image);
      const { imageAlternativeText, imageUrl } = props;

      expectComponentToHaveProps(wrapper, {
        alt: imageAlternativeText,
        src: imageUrl,
      });
    });
  });

  describe('the `Card.Content` component', () => {
    it('should render the right children', () => {
      const wrapper = getFeaturedRoomType().find(Card.Content);

      expectComponentToHaveChildren(
        wrapper,
        Card.Header,
        Card.Description,
        Card.Description,
        Card.Description
      );
    });
  });

  describe('the `Card.Header` component', () => {
    it('should have the right `children`', () => {
      const wrapper = getFeaturedRoomType().find(Card.Header);

      expectComponentToHaveChildren(wrapper, props.roomTypeName);
    });
  });

  describe('the first `Card.Description` component', () => {
    it('should have the right `children`', () => {
      const wrapper = getFeaturedRoomType()
        .find(Card.Description)
        .at(0);

      expectComponentToHaveChildren(wrapper, props.locationName);
    });
  });

  describe('the second `Card.Description` component', () => {
    it('should have the right `children`', () => {
      const wrapper = getFeaturedRoomType()
        .find(Card.Description)
        .at(1);

      expectComponentToHaveChildren(
        wrapper,
        getRoomTypeDescription(props.guestsNumber, props.bedsNumber)
      );
    });
  });

  describe('the third `Card.Description` component', () => {
    it('should have the right `children`', () => {
      const wrapper = getFeaturedRoomType()
        .find(Card.Description)
        .at(2)
        .children();

      expectComponentToBe(wrapper, 'span');
    });
  });

  describe('the span', () => {
    it('should have the right children', () => {
      const wrapper = getFeaturedRoomType().find('span');

      expectComponentToHaveChildren(wrapper, 'from ', Heading, '/night');
    });
  });

  describe('the `Heading` component', () => {
    const getHeading = () => getFeaturedRoomType().find(Heading);

    it('should have the right props', () => {
      const wrapper = getHeading();

      expectComponentToHaveProps(wrapper, { size: 'medium' });
    });

    it('should have the right children', () => {
      const wrapper = getHeading();

      expectComponentToHaveChildren(wrapper, props.nightPrice);
    });
  });

  it('should have `displayName` `FeaturedRoomType`', () => {
    expectComponentToHaveDisplayName(FeaturedRoomType, 'FeaturedRoomType');
  });
});
