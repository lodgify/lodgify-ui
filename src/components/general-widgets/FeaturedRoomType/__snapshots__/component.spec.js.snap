// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FeaturedRoomType /> should render the right structure 1`] = `
<Card
  className="has-featured"
  href="/"
>
  <ResponsiveImage
    alternativeText="this alt tag"
    hasRoundedCorners={false}
    imageHeight={null}
    imageNotFoundLabelText="Image not found!"
    imageTitle="Image Title"
    imageUrl="🐱🐱"
    imageWidth={null}
    isAvatar={false}
    isCircular={false}
    isFluid={true}
    label={null}
    sizes={null}
    srcSet={null}
  />
  <CardContent>
    <CardHeader>
      The Cat House
    </CardHeader>
    <CardDescription>
      Catania
    </CardDescription>
    <CardDescription>
      🙆‍♂️: 3 | 🛏: 3
    </CardDescription>
    <CardDescription>
      <span>
        from 
        <Heading
          isColorInverted={false}
          size="medium"
          textAlign={null}
        >
          $280
        </Heading>
        /night
      </span>
    </CardDescription>
  </CardContent>
</Card>
`;
